### DISCUSSION TITLE
ON_COORD_SET FLIP

### DISCUSSION THREAD

**Discussion by amendolajamie** [indilib/indi#1875] [General]
I would like to add a "FLIP" member to ON_COORD_SET to allow mount flip commands to be more explicit. Losmandy mounts have commands that are used for flipping that are more reliable than goto commands, and having a ON_COORD_SET member that makes this explicit will allow these commands to be used reliably. I can make the inditelescope flip command default to issuing a goto if flip is not implemented in the telescope driver to leave existing drivers unaffected.

**Comment by knro** [comment]
This has to be done via a separate property. ON_COORD_SET is a stable property that many clients rely on as-is.

**Comment by amendolajamie** [comment]
I added a flip property that is backwards compatible, and only added if the TELESCOPE_CAN_FLIP capability is set in the driver. Ekos will check for the property and use it if it is available. I don't think that it will have any adverse affects on existing use cases. Please review.

https://github.com/indilib/indi/pull/1880
https://invent.kde.org/education/kstars/-/merge_requests/857

**Comment by knro** [comment]
So without this, how are they flipping now? is there a limit set after which any GOTO command will trigger a flip?

**Comment by amendolajamie** [comment]
There is a "western limit" that the go-to command uses to determine if it should flip. It works most of the time, but could fail under some conditions. It was designed for visual use and depending on the destination coordinates, safety and goto limit combinations, may not perform a flip.

This update allows the Gemini to use the "MM" command. This one always flips if it is within safety limits. It does not use the "western limits", so it does not  require the user to convert units and enter western
go-to limits in the mount.

It will also allow other mounts to use their own specialized flip command
if it is available, and makes sense for that mount.

**Comment by knro** [comment]
Ok this should be considered in the next INDI development cycle as we're too close to release now.

**Comment by sterne-jaeger** [comment]
It seems like mount manufacturers have different solutions for this. Avalon for example has a dedicated meridian flip setting with the three options disabled + auto + forced. Therefore we should think about it twice before we add a mount specific behavior to the INDI standard mount.

What is exactly the semantics of the FLIP setting on Losmandy mounts? Does it always try to switch the pier side?

**Reply by amendolajamie** [reply]
The :MM# command will do a meridian unless the flip will result in going past the safety limits.

**Reply by sterne-jaeger** [reply]
That means, as long as the safety limits allow it, twice sending :MM# results in flipping back and forth?

**Comment by amendolajamie** [comment]
That's correct. It will flip every time it's called unless the flip will be outside safety limits.

**Reply by sterne-jaeger** [reply]
I'm still quite reluctant adding this behavior to other scopes than Losmandy scopes. I did a quick check for popular scope manufacturers like Skywatcher or Ioptron. Both do not have the Losmandy behavior implemented.

Additionally, we should be careful using this setting for automatic meridian flips like it is implemented in KStars/EKOS. What happens with this switch when the mount has completed its meridian flip? Does it automatically change to TRACK or SLEW? Otherwise, it could be that the next slew command will cause another change of the pier side - which might be unintended.

@amendolajamie There is a GOTO limit setting in the Losmandy Gemini setup that exactly supports the automatic meridian flip behavior. If it is for example set to 90, all slews to a position western of the meridian will lead to an automatic meridian flip. With this setting I would expect that meridian flips with KStars/EKOS should work correctly.

**Reply by amendolajamie** [reply]
For the losmandy mount, the western limits + slew command usually works, but there are combinations of position + limit + go-to destination that will fail. For this reason, the ascom driver also uses the ":MM#" command to preform a flip. I'll review the other mounts to see which ones would benefit from a specific flip command.

**Reply by sterne-jaeger** [reply]
I don't disagree adding this behavior to the Losmandy driver, but please make it Losmandy specific without touching the general LX200 driver, since this will have side effects on others.

**Comment by amendolajamie** [comment]
I found these methods that mounts use to configure / determine if a flip should be initiated:

- LX200_10MICRON::flip() :FLIP#
-- For the german eq mount this appears to have a similar behavior to the Losmandy flip where there is a specific command for flipping. 

- Avalon, and Synscan devices seem to have a "Forced" meridian flip, where you set the forced flag, then issue a goto. Although this is not exactly the same, it is similar.

- Others seem to use specialized limits, some that can be set by the user to determine if a goto will initiate a flip.

It seems that there is a handful of methods that mounts use to determine if a flip will be initiated. It does seem to me that a "FLIP" property makes sense since this is something that most, if not all GEM mounts manage in some form or other. In the case of a "forced flip" the forced command would precede the goto. In the case of a user configurable, or built-in limit, the "FLIP" property could return a warning or error if the limit will not allow a flip to occur. 

Jamie

**Reply by sterne-jaeger** [reply]
I think the main question is whether we could generalize it such that we could use it for the automatic meridian flip of EKOS. And here my concern is that we should be very careful with forcing flips. If possible I would prefer that we leave the decision of the pier side to the mount.
Are you really sure that there is no way to configure Gemini in such a way that the automatic flip takes always place? Could you give concrete examples where this fails although the GOTO limit is set?

**Reply by amendolajamie** [reply]
I learned about the failures through discussions with the Losmandy support and developers. They requested that I update indi to use the "MM" command to reduce failures that people experience from using the regular go-to with a western limits. They updated the ascom driver to use the flip specific command because it is more reliable. To me, the addition of a flip property seems like it is a good way of encapsulating the flip behaviors of different mounts. It can default to a regular go-to, and it could do something specific if for mounts that do have flip commands. I will of-course defer to the main indi developers/owners decision.

**Reply by sterne-jaeger** [reply]
Under which conditions does the ASCOM driver send the MM command? Does it have a specific "Flip" command? And how is this used by typical ASCOM clients?

Please keep in mind, in EKOS we are assuming the mount has an automatic flip mode and do not touch this setting. I think if we want to use the MM command, we need to do some temporary shift to a "force flip" mode - but only when we want to trigger a flip. And afterwards, we need to reset it to the former state. Possible, but might become tricky...

**Comment by amendolajamie** [comment]
Ascom does have a side-of-pier property that can be writable if the mount
supports it.

Please check my implementation. It just adds a flip property to indi and a
capability flag. If the flip capability is set, then the flip will be used
and the mount specific driver will handle the flip in whatever way is
required by the mount. If the flip capability is not set then it defaults
to a regular slew command. This will allow all mounts to remain the same.
The Gemini driver will use the MM command to flip. Other mounts that have a
"force flip" feature can be updated to use their specific flip in whatever
manner they require.

Jamie

**Reply by sterne-jaeger** [reply]
I commented already on your PR here. Please restrict your changes to the Gemini classes - at least for a first step.

**Reply by amendolajamie** [reply]
OH, sure I'll do that.

**Reply by amendolajamie** [reply]
I've moved the changes to ls200gemini. Please check https://github.com/indilib/indi/pull/1880
